### PR TITLE
support workspace resource quota

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

`resourcequotas.quota.kubesphere.io` mirrors ResourceQuota at a cluster scope.  This object is easily convertible to synthetic ResourceQuota object to allow quota evaluation re-use.

**Which issue(s) this PR fixes**:

Related to https://github.com/kubesphere/kubesphere/issues/2939

**Special notes for reviewers**:

The main idea comes from https://github.com/openshift/api/blob/master/quota/v1/types.go

1. Add `resourcequotas.quota.kubesphere.io` CRD.
2. Implement `resourcesquotas.quota.kubesphere.io` admission webhook.

**Additional documentation, usage docs, etc.**:

Workspace resource quota can be defined across multiple projects. Resources used in the selected workspace will be aggregated and that aggregate will be used to limit resources across all the projects in the selected workspace.

image

### CODE REVIEW

**Comment by zryfish** [comment]
How is this PR going? Any update?

File: pkg/controller/quota/resourcequota_controller.go [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+	recorder                record.EventRecorder
+	maxConcurrentReconciles int
+	// Knows how to calculate usage
+	registry quotav1.Registry
+	// Controls full recalculation of quota usage
+	resyncPeriod time.Duration
+	scheme       *runtime.Scheme
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, maxConcurrentReconciles int, resyncPeriod time.Duration, informerFactory k8sinformers.SharedInformerFactory) error {
```
Comment (RolandMa1986) [inline pkg/controller/quota/resourcequota_controller.go:78]: Should the quota Reconciler be setup in _server.go_?

File: pkg/apis/quota/group.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
+Copyright 2019 The KubeSphere Authors.
```
Comment (RolandMa1986) [inline pkg/apis/quota/group.go:2]: Copyright should be consistent in the commits.
```suggestion
Copyright 2021 The KubeSphere Authors.
```

**Review by RolandMa1986 — CHANGES_REQUESTED** [review]
This is a handy commit in general. But seems I have lost some context. It would be great if there's a design document for the reviewers.

File: kube/pkg/api/service/util.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/*
```
[the remark this answers is not in the document]
Comment (wansir) [reply kube/pkg/api/service/util.go]: fixed

File: pkg/controller/quota/resourcequota_controller.go (hunk 1 above)
Comment (wansir) [reply pkg/controller/quota/resourcequota_controller.go:78]: fixed

File: pkg/apis/quota/group.go (hunk 2 above)
Comment (wansir) [reply pkg/apis/quota/group.go]: fixed

File: vendor/kubesphere.io/client-go [hunk 4]
```diff
@@ -1 +0,0 @@
-../../staging/src/kubesphere.io/client-go
```
[the remark this answers is not in the document]
Comment (wansir) [reply vendor/kubesphere.io/client-go:1]: Conflict with the master branch. `update-vendor.sh` needs to be updated.

File: pkg/controller/quota/resourcequota_controller.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+		actualUsage, err := quotaUsageCalculationFunc(namespaceName, quota.Spec.Quota.Scopes, quota.Spec.Quota.Hard, r.registry, quota.Spec.Quota.ScopeSelector)
+		if err != nil {
+			return err
+		}
+		recalculatedStatus := corev1.ResourceQuotaStatus{
+			Used: actualUsage,
+			Hard: quota.Spec.Quota.Hard,
+		}
+
+		// subtract old usage, add new usage
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/controller/quota/resourcequota_controller.go:269]: `Total.Used` needs to be recalculated after namespace resource usage changes. The code is copied from https://github.com/openshift/cluster-policy-controller/blob/release-4.9/pkg/quota/clusterquotareconciliation/reconciliation_controller.go#L357-L363.

File: pkg/controller/quota/resourcequota_controller.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+	matchingNamespaceNames := make([]string, 0)
+	for _, namespace := range matchingNamespaceList.Items {
+		matchingNamespaceNames = append(matchingNamespaceNames, namespace.Name)
+	}
+
+	for _, namespace := range matchingNamespaceList.Items {
+		namespaceName := namespace.Name
+		namespaceTotals, namespaceLoaded := getResourceQuotasStatusByNamespace(quota.Status.Namespaces, namespaceName)
+		if !sliceutil.HasString(matchingNamespaceNames, namespaceName) && namespaceLoaded {
```
[the remark this answers is not in the document]
Comment (wansir) [reply pkg/controller/quota/resourcequota_controller.go]: You are right, useless code removed.

File: pkg/controller/quota/resourcequota_controller.go [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+const (
+	ControllerName                 = "resourcequota-controller"
+	DefaultResyncPeriod            = 5 * time.Minute
+	DefaultMaxConcurrentReconciles = 8
+)
+
+// Reconciler reconciles a Workspace object
+type Reconciler struct {
+	client.Client
+	logger                  logr.Logger
```
Comment (zryfish) [inline pkg/controller/quota/resourcequota_controller.go:68]: Please use `klog` instead of `logr`, there is an alternative in klog `klog.logr`

Comment (wansir) [reply pkg/controller/quota/resourcequota_controller.go:68]: `klog.klogr` is used indirectly, controler-runtime provides global logger configuration. See this:
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.4/pkg/log/log.go#L46-L73
https://github.com/kubesphere/kubesphere/blob/master/cmd/controller-manager/app/server.go#L186

**Pull request merged**